history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
```

Comment:
I'm very sorry for chiming in very late here. I would like to clarify a few more things.

> At t == 0, we don't have any STRs, we've just handed out a few temporary bindings. We can set them in the tree, but since we haven't signed the root yet, any lookups just come from the temporary bindings.

Whether we start counting epochs at t=0 or t=-1, the initial epoch *does* need an STR, even if the tree is empty, since we need to return a proof of absence along with any TB.

>> At t == 1, we take a snapshot (clone) of the tree as it is and sign the root.
> So this STR is the STR for the tree at ``t == 1``. The temporary bindings (namely ``TB_0``) which is set in the tree at ``t == 1`` (namely ``tree_1``) would belong to the tree at ``t == 0`` (namely ``tree_0``), since we inserted them into the tree before cloning the tree.
> 
> As a result, the ``STR[0]`` of ``tree_0`` (or ``array[0]`` as you called) is not computed based on ``TB_0``, whereas the ``STR[1]`` of ``tree_1`` would be computed based on ``TB_0``. 
> Is that right?

Yeah, this is right. But even if we included all ``TB_0``s in a batch at the end of t=0 like you had originally suggested (and as we do in the reference implementation), since the initial epoch t=0 has an STR,  the next STR for t=1 and ``tree_1`` would be computed based on all ``TB_0``s.